### PR TITLE
[pulsar-broker] add uniform load shedder strategy to distribute traffic uniformly across brokers

### PR DESCRIPTION
**Pull request by rdhabalia**
### Motivation
Pulsar Broker provides a [configurable way](https://github.com/apache/pulsar/blob/master/pulsar-broker-common/src/main/java/org/apache/pulsar/broker/ServiceConfiguration.java#L1711) to configure various load shedding strategies based on usecases. It already has strategies to load balance based on broker resource utilization ([ThresholdShedder](https://github.com/apache/pulsar/blob/master/pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/impl/ThresholdShedder.java)) or based on bundle usage threshold ([OverloadShedder](ache/pulsar/broker/loadbalance/impl/OverloadShedder.java)). 
We need another strategy to distribute load uniformly across all brokers. So, we are adding `UniformLoadShedder` strategy to achieve goal for uniform load in cluster.

### NOTE
`UniformLoadShedder` also requires isolation and broker weight mechanism to also use strategy in heterogeneous hardware.  I will create another PR for this enhancement.

### CODE REVIEW

**Review by eolivelli — APPROVED** [review]
+1
But I am not expert of this kind of algorithms.
So it is better to wait for some other review

**Comment by codelipenghui** [comment]
@hangc0276 Please help review the PR.

File: conf/broker.conf
Comment (Jason918) [inline conf/broker.conf:1104]: IMHO, It's better to add a notice about how to enable uniform load shedding, like set `loadBalancerLoadSheddingStrategy="org.apache.pulsar.broker.loadbalance.impl.UniformLoadShedder"`

**Review by Jason918 — COMMENTED** [review]
And we need to update the doc of `loadBalancerLoadSheddingStrategy` about this new strategy.

File: pulsar-broker-common/src/main/java/org/apache/pulsar/broker/ServiceConfiguration.java
Comment (lhotari) [inline pulsar-broker-common/src/main/java/org/apache/pulsar/broker/ServiceConfiguration.java:1793]: just wondering whether there would there need to be support for fractions like 1.5 ?

File: conf/broker.conf
Comment (Anonymitaet) [reply conf/broker.conf:1104]: +1

@rdhabalia 
For the doc side, how about adding docs like below? (feel free to correct me if my understanding is wrong)
image
That is:
- Add a section `Shed load manually` 
- 3 strategies (`ThresholdShedder`, `OverloadShedder`, and `UniformLoadShedder`) belong to `Shed load manually` section
- Add explanations of "when to use" and "how to use" for each strategy

If so, can you add them? Thanks!

**Review by Anonymitaet — CHANGES_REQUESTED** [review]

**Review by Anonymitaet — COMMENTED** [review]
@rdhabalia For the doc side, how about adding docs like below:

- Add a section `Load shedder strategy` to [load balance](https://pulsar.apache.org/docs/en/next/administration-load-balance/) chapter
- Explain 3 strategies (`ThresholdShedder`, `OverloadShedder`, and `UniformLoadShedder`), including "when to use" and "how to use" for each strategy

If so, can you add them? Thanks!

File: pulsar-broker-common/src/main/java/org/apache/pulsar/broker/ServiceConfiguration.java
Comment (rdhabalia) [reply pulsar-broker-common/src/main/java/org/apache/pulsar/broker/ServiceConfiguration.java]: yes, value is double and it supports fractions.

File: conf/broker.conf
Comment (rdhabalia) [reply conf/broker.conf:1104]: added into documentation

**Comment by rdhabalia** [comment]
@Anonymitaet I have added documentation, if needed then we can always add more documentations in future.

**Comment by codelipenghui** [comment]
I think we'd better use the batch rate instead of the msg rate? the batch rate consumes the broker CPU resource and throughput consumes the network bandwidth.

**Review by Anonymitaet — CHANGES_REQUESTED** [review]
Hi @rdhabalia thanks for adding docs, but users may not have enough patience to wade through “chunks" of information, so I re-organize them [here](https://docs.google.com/document/d/1kZvYreXMFPXBUtU521O-Dlsh2MGln7-ZlrcJNH0aoHI/edit) and left some comments. 

Can you help review and answer the questions? If the content makes sense, you can implement it into your .md files, thanks.

**Comment by rdhabalia** [comment]
@codelipenghui 

> I think we'd better use the batch rate instead of the msg rate?

LoadBalancer only captures `MsgRate` and not batch msg, and that's because  batch message is an abstraction for the client. Broker still considers batch message a single message in terms of processing and storage. batch-message size which something matters and that can be covered by msgThroughput threshold.

@Anonymitaet 

Please don't block the PR due to the documentation format. yes, I agree, the better format helps but the reorganization of the content can be handled in separate PR as well because every individual has a different idea about the formation.

**Comment by rdhabalia** [comment]
@Anonymitaet fixed typo about `loadBalancerMsgRateDifferenceShedderThreshold` config based on your feedback in the [doc](https://docs.google.com/document/d/1kZvYreXMFPXBUtU521O-Dlsh2MGln7-ZlrcJNH0aoHI/edit)

**Comment by Anonymitaet** [comment]
> @Anonymitaet
> 
> Please don't block the PR due to the documentation format. yes, I agree, the better format helps but the reorganization of the content can be handled in separate PR as well because every individual has a different idea about the formation.

Hi @rdhabalia, 
- I did not mean to block anything, I just want to improve the user experience since no one would like to read "a mountain of" information at a glance. Technical information should be easy to find/understand/use.
- Agree that documentation can be submitted in a separate PR. 
- Everyone has a different opinion on everything but there are some general rules/standards accepted by the majority. In technical writing, often, text that is difficult to understand in paragraph form becomes clear in a table format. Tables help highlight relationships among similar pieces of information and are easily scanned for quick reference.

**Comment by rdhabalia** [comment]
@Anonymitaet sure, I agree. let me create a separate PR with detailed info because we still have to add support of namespace isolation. so, can we unblock this PR and merge it for now and we can enhance with better user experience.

**Comment by Anonymitaet** [comment]
@rdhabalia yes, it makes sense. Separate doc PR is more convenient for cutting a release (cherry-pick changes) as well.

**Comment by rdhabalia** [comment]
@Anonymitaet I still see blocker for this PR. Please avoid creating blocker for DOC.

**Comment by Anonymitaet** [comment]
@rdhabalia I label this PR with `doc-required`. Do not forget to revert doc changes in this PR and submit them in a separate PR, thanks.

**Pull request merged**
